manager/state/raft/storage.go

```diff
@@ -274,25 +275,106 @@ func (n *Node) saveSnapshot(snapshot raftpb.Snapshot, keepOldSnapshots uint64) e
 	// This means that if the current snapshot doesn't appear in the
 	// directory for some strange reason, we won't delete anything, which
 	// is the safe behavior.
+	curSnapshotIdx := -1
 	var (
-		afterCurSnapshot bool
-		removeErr        error
+		removeErr          error
+		remainingSnapshots []string
 	)
+
 	for i, snapFile := range snapshots {
-		if afterCurSnapshot {
-			if uint64(len(snapshots)-i) <= keepOldSnapshots {
-				return removeErr
-			}
-			err := os.Remove(filepath.Join(n.snapDir(), snapFile))
-			if err != nil && removeErr == nil {
-				removeErr = err
+		if curSnapshotIdx >= 0 && i > curSnapshotIdx {
+			if uint64(i-curSnapshotIdx) > keepOldSnapshots {
+				err := os.Remove(filepath.Join(n.snapDir(), snapFile))
+				if err != nil && removeErr == nil {
+					removeErr = err
+				}
+				continue
 			}
 		} else if snapFile == curSnapshot {
-			afterCurSnapshot = true
+			curSnapshotIdx = i
+		}
+		remainingSnapshots = append(remainingSnapshots, snapFile)
+	}
+
+	if removeErr != nil {
+		return removeErr
+	}
+
+	// Remove any WAL files that only contain data from before the oldest
+	// remaining snapshot.
+
+	if len(remainingSnapshots) == 0 {
+		return nil
+	}
+
+	// Sort snapshot filenames in lexical order
+	sort.Sort(sort.StringSlice(remainingSnapshots))
```

Comment:
`remainingSnapshots` is already sorted, (in reversed order), because `snapshots` is sorted.

We may even replace this array with `var oldestSnap string` and construct it from the loop above.